engine/resources/hetzner.go

```diff
@@ -0,0 +1,716 @@
+// requests a server deletion from the API, and waits until confirmed.
+// TODO: add deleteServer permission field
+func (obj *HetznerRes) deleteServer(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("deleteServer()")
+	}
+	if obj.server == nil {
+		return fmt.Errorf("server is already unavailable")
+	}
+	if _, err := obj.client.Server.Delete(ctx, obj.server); err != nil {
+		return errwrap.Wrapf(err, "client.server.delete failed")
+	}
+	if obj.init.Debug {
+		obj.init.Logf("waiting until deletion is confirmed")
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("wait interrupted by context")
+		case <-time.After(time.Duration(obj.WaitInterval) * time.Second):
+			server, _, err := obj.client.Server.GetByName(ctx, obj.Name())
+			if err != nil {
+				return errwrap.Wrapf(err, "failed serverupdate request")
+			}
+			if server == nil {
+				obj.init.Logf("server deletion confirmed")
+				obj.server = server
+				return nil
+			}
+		}
+	}
+}
+
+// rebuildServer deletes the current server instance and creates a new one,
+// in accordance with the provided resource specifications.
+// TODO: add rebuild permissions! (or split into create & delete permissions)
+func (obj *HetznerRes) rebuildServer(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("rebuildServer()")
+	}
+	if obj.server == nil {
+		return fmt.Errorf("server is unavailable")
+	}
+	if err := obj.deleteServer(ctx); err != nil {
+		return errwrap.Wrapf(err, "deleteServer failed")
+	}
+	if err := obj.createServer(ctx); err != nil {
+		return errwrap.Wrapf(err, "createServer failed")
+	}
+	return nil
+}
+
+// getServerConfig converts resource fields to hcloud.ServerCreateOpts struct,
+// such that a server instance created with those Opts has the desired specs.
+func (obj *HetznerRes) getServerConfig(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("getServerConfig()")
+	}
+	// default, volumes not supported (yet)
+	automount := false
+	// direct poweron to speed things up a little
+	startAfterCreate := (obj.State == StateRunning)
+	// build hcloud-typed specification structs
+	serverType, _, err := obj.client.ServerType.GetByName(ctx, obj.Servertype)
+	if err != nil {
+		return errwrap.Wrapf(err, "failed to collect Servertype struct")
+	}
+	image, _, err := obj.client.Image.GetByName(ctx, obj.Image)
+	if err != nil {
+		return errwrap.Wrapf(err, "failed to collect Image struct")
+	}
+	datacenter, _, err := obj.client.Datacenter.GetByName(ctx, obj.Datacenter)
+	if err != nil {
+		return errwrap.Wrapf(err, "failed to collect Datacenter struct")
+	}
+	// TODO: add more flexible key selection
+	keylist, err := obj.client.SSHKey.All(ctx)
+	if err != nil {
+		return errwrap.Wrapf(err, "failed to collect SSHKey array")
+	}
+	// build serverconfig from given specs & defaults
+	obj.serverconfig = hcloud.ServerCreateOpts{
+		Name:             obj.Name(),        //string
```

Comment:
if you want to leave comments with all the types, leave a space after the // so that nobody thinks they're magic strings and so it's easier to read.